plugins/cpu/cpu.go

```diff
@@ -13,31 +21,170 @@ const (
 	niceGauge
 )
 
+// Hold cpu sampled data and calculate moving averages
+type sample struct {
+	alpha        float64
+	values       []float64
+	expMovingAvg float64
+}
+
+func newSample(alpha float64) *sample {
+	return &sample{alpha: alpha}
+}
+
+func (s *sample) add(v float64) {
+	s.values = append(s.values, v)
+
+	if s.expMovingAvg == 0 {
+		s.expMovingAvg = v
+	} else {
+		s.expMovingAvg = v*s.alpha + (s.expMovingAvg * (1 - s.alpha))
+	}
+}
+
+// Reset data at reporting intervals
+func (s *sample) drain() float64 {
+	avg := s.expMovingAvg
+	s.expMovingAvg = 0
+	s.values = []float64{}
+	return avg
+}
+
+func (s *sample) movingAvg() float64 {
+	if len(s.values) == 0 {
+		return 0
+	}
+	var (
+		subSet     = (len(s.values) / 2) + (len(s.values) % 2)
+		avgs       []float64
+		sum, count float64
+	)
+
+	for i := 0; i <= len(s.values)/2; i++ {
+		for _, v := range s.values[i : i+subSet] {
+			sum += v
+			count += 1
+		}
+		avgs = append(avgs, (sum / count))
+		sum = 0
+		count = 0
+	}
+
+	for _, v := range avgs {
+		sum += v
+		count += 1
+	}
+
+	return sum / count
+}
+
 type CPU struct {
+	mux           sync.Mutex
 	previous      []int
 	current       []int
 	currentTotal  int
 	previousTotal int
 	lastUpdate    time.Time
+	averages      []*sample
+	updaters      []metrics.Updater
+	slackURL      string
+}
 
-	updaters []metrics.Updater
+type Config struct {
+	Threshold         float64
+	SampleRate        time.Duration
+	ReportingInterval time.Duration
+	SlackURL          string
 }
 
-func New() *CPU {
-	return &CPU{
+func New(cfg *Config) *CPU {
+	// Determine the alpha factor for exponential moving average
+	alpha := cfg.SampleRate.Seconds() / cfg.ReportingInterval.Seconds()
```

Comment:
A more useful comment here would be how we arrived at this as a good value for alpha.